### ISSUE TITLE
Consider Removal of remotestate.Backend

### ISSUE THREAD

**Issue by hasheddan** [hashicorp/terraform#21744]
### Current Terraform Version

```
v0.12.2
```

### Use-cases
Terraform remote-state backends currently all implement the `backend.Backend` interface by embedding the `schema.Backend` struct. In the early days of TF, `Consul` (and later `Inmem`) used `remotestate.Backend` to serve the same purpose that `schema.Backend` now provides. Therefore, `remotestate.Backend` is now obsolete and functionality will not be reduced in any manner if it is removed.

### Attempted Solutions
N/A - issue does not currently impede Terraform functionality. It is only a code clarity improvement.

### Proposal
Removal of `remotestate.Backend`.
This would include removing:
- [remote-state/backend.go](https://github.com/hashicorp/terraform/blob/master/backend/remote-state/backend.go)
- [remote-state/backend_test.go](https://github.com/hashicorp/terraform/blob/master/backend/remote-state/backend_test.go)
- [remote-state/testing.go](https://github.com/hashicorp/terraform/blob/master/backend/remote-state/testing.go)

I would be more than happy to open a PR to accomplish this.

### References
See reference to old version of `Consul` remote-state backend that used `remotestate.Backend` instead of `schema.Backend` [here](https://github.com/hashicorp/terraform/blob/1f5d425428b061c2e90f2e0092f1729332e00cbc/backend/remote-state/consul/backend.go).

**Comment by apparentlymart** [comment]
Hi @hasheddan! Thanks for raising this.

We have some longer-term plans to refactor the backend code so that the separations between them are better suited to how the system evolved in practice, particularly with the addition of the new "remote" backend which both works differently and currently confusingly overlaps in name with the "remote state" parts of the codebase, including what you referred to here.

We've not yet done the full design work to see what this new model might look like, and so we'd been leaving the backend code largely as-is until we have the time to dig into that more.

With that said, if you have seen that removing these files can be done without any significant changes elsewhere, I think we'd be open to it. We don't want to make any _risky_ changes to these packages until we understand better what the new architecture will look like, but if it's just dead code cleanup then indeed there's little reason not to do it!

**Comment by hasheddan** [comment]
Thanks for following up on this @apparentlymart! Really appreciate all the work you do in the HashiCorp open source community :) I have removed the specified files and unit tests still work as expected. Furthermore, I have checked to make sure that there are no references to `remotestate.Backend` anywhere in the code base, so I believe that it will be okay to go ahead and remove them. I will open a PR for this to be reviewed if it sounds good to you!
